File: sdks/python/replayai/steps.py
"""Step recording.

``record_step()`` appends a step to the current session (the innermost
:class:`~replayai.context.TraceContext` on the ContextVar stack).
``arecord_step()`` is the async variant — same behavior, awaitable so it
can be interleaved with async agent code without blocking the event loop.
"""
from __future__ import annotations

import asyncio
from typing import Any

from .context import TraceContext, _append_step, current_session


def record_step(**step: Any) -> None:
    """Record a step into the current trace.

    Keyword arguments (all optional except where noted):

    - ``type`` (str, default ``"llm_call"``) — one of
      ``llm_call``, ``tool_call``, ``retrieval``, ``decision``, ``error``.
    - ``name`` (str) — step label shown in the timeline.
    - ``model`` (str) — model name, used for cost estimation.
    - ``tokens_in`` / ``tokensIn`` (int) — prompt tokens.
    - ``tokens_out`` / ``tokensOut`` (int) — completion tokens.
    - ``input`` (str|dict) — step input (redacted before persist).
    - ``output`` (str|dict) — step output (redacted before persist).
    - ``status`` (str, default ``"success"``) — ``success`` | ``failed`` | ``running``.
    - ``offset_ms`` / ``offsetMs`` / ``t`` (int) — explicit offset from
      session start. Defaults to wall-clock since session start.
    - ``duration_ms`` / ``durationMs`` (int) — explicit step duration.
    """
    ctx = current_session()
    if ctx is None or ctx.session is None:
        # No active trace — silently drop unless strict.
        from . import config as _config

        if _config.strict_mode:
            raise RuntimeError(
                "record_step() called outside of an active trace() context."
            )
        return
    _append_step(ctx.session, step)


async def arecord_step(**step: Any) -> None:
    """Async variant of :func:`record_step`.

    Identical behavior — awaits a 0-delay yield first so callers can use
    ``await arecord_step(...)`` symmetrically inside async agent code.
    """
    # Yield control so the call signature stays awaitable even when there's
    # no real async work to do.
    await asyncio.sleep(0)
    record_step(**step)

File: sdks/typescript/src/steps.ts
// ReplayAI TypeScript SDK — manual step recording.

import { appendStep } from "./context.js";
import { redactText } from "./redact.js";
import type { RecordStepInput, SessionStep, StepStatus, StepType } from "./types.js";

const VALID_TYPES: StepType[] = ["llm_call", "tool_call", "retrieval", "decision", "error"];
const VALID_STATUSES: StepStatus[] = ["success", "failed", "running", "warning"];

function coerceType(v: string | undefined): StepType {
  if (v && (VALID_TYPES as string[]).includes(v)) return v as StepType;
  return "llm_call";
}

function coerceStatus(v: string | undefined): StepStatus {
  if (v && (VALID_STATUSES as string[]).includes(v)) return v as StepStatus;
  return "success";
}

/**
 * Record a step into the current trace session.
 *
 * No-op when called outside a `withTrace()`/`trace()` context (unless strict
 * mode is on, in which case it warns). Inputs and outputs are redacted with
 * the configured patterns before persistence.
 */
export async function recordStep(input: RecordStepInput): Promise<void> {
  const step: SessionStep = {
    type: coerceType(input.type),
    name: input.name,
    model: input.model ?? null,
    tokensIn: input.tokensIn ?? 0,
    tokensOut: input.tokensOut ?? 0,
    input: redactText(input.input),
    output: redactText(input.output),
    status: coerceStatus(input.status),
    t: input.t ?? input.offsetMs,
    offsetMs: input.offsetMs ?? input.t,
    durationMs: input.durationMs ?? 0,
  };
  appendStep(step);
}

/** Sync variant — same as `recordStep()` but without the await ceremony. */
export function recordStepSync(input: RecordStepInput): void {
  const step: SessionStep = {
    type: coerceType(input.type),
    name: input.name,
    model: input.model ?? null,
    tokensIn: input.tokensIn ?? 0,
    tokensOut: input.tokensOut ?? 0,
    input: redactText(input.input),
    output: redactText(input.output),
    status: coerceStatus(input.status),
    t: input.t ?? input.offsetMs,
    offsetMs: input.offsetMs ?? input.t,
    durationMs: input.durationMs ?? 0,
  };
  appendStep(step);
}
